open-hackathon-server/src/hackathon/expr/expr_mgr.py

```diff
@@ -254,10 +222,37 @@ def __start_new_expr(self, hackathon, template, user):
         context = starter.start_expr(Context(
             template=template,
             user=user,
-            hackathon=hackathon))
+            hackathon=hackathon,
+            pre_alloc_enabled = False))
 
         return self.__report_expr_status(context.experiment)
 
+    def start_pre_alloc_exprs(self, user, template_name, hackathon_name=None, pre_alloc_num = 0):
+        self.log.debug("start_pre_alloc_exprs: %d " % pre_alloc_num)
+        if pre_alloc_num == 0: return
+
+        hackathon = self.__verify_hackathon(hackathon_name)
+        template = self.__verify_template(hackathon, template_name)
+
+
+        starter = self.get_starter(hackathon, template)
+        if not starter:
+            raise PreconditionFailed("either template not supported or hackathon resource not configured")
+
+        while pre_alloc_num > 0:
```

Comment:
We have to check how many pre-allocated expr are already there. Otherwise, since `start_pre_alloc_exprs` is triggered in cron job, it will start 100 expr everytime the cron job runs.

Reply:
I thought through it, it's better not be terminated during the progress and it must be kept simple and clear, otherwise, to many dependencies(like the data in experiment table, their status, which one should be assigned,......)
what i expect is:
1, disable pre_alloc in config.py
2,start the service (run.py)
3,create template and hackathon
4,terminate the service
5,enable pre_alloc and config pre_allocate_num
6,restart service and allocated the resource as the config
7, let users in when everything is ready (kube may not be able to assign resource as many as we apply, the pre allocate pregress will be terminated once it's failed, accept as many as kube assigns us, that's the logic here)

Reply:
那需要人时刻盯着。。。。需要经常改配置,不太好吧。。。

Reply:
我们肯定得盯着，特别是前几次，稳定了可能还好些

Reply:
这个等于说先起几个，然后关掉。等预分配的用掉了，又打开，然后又关掉。这个实际运维的时候不可取啊。。。要靠程序来监督，不是人来监督，比如刚好人不在呢? 这个计划任务20分钟就跑一次，人稍微反应满了，就又是一批呀。。。

Reply:
这样，咱们先这么跑一次，等优麒麟的活动结束了，再改。我们就先盯着点，等这个单一的过程没问题了，我们把add_once改成add_interval，然后把experiment的数据理顺，什么样的experiment是可以分配的，什么是已经被占用的，然后看看还剩多少，继续分配。我觉得这样比较稳妥。

Reply:
而且我觉得20分钟起一次也不是一个好的选择，应该由运维人员手动触发，放到管理页面，比如我在119测试的时候，启动7个之后就没有cpu资源了，这样还在后台每20分钟就给kube发一次请求，也得不到想要的。

Reply:
手动触发是肯定不行的，不可能有运维时刻盯着

Reply:
我们都没有运维。。。。